kylin-spark-project/kylin-spark-engine/src/main/scala/org/apache/kylin/engine/spark/builder/CubeTableEncoder.scala

```diff
@@ -61,15 +66,59 @@ object CubeTableEncoder extends Logging {
         val encodeColRef = convertFromDot(ref.identity)
         val columnIndex = structType.fieldIndex(encodeColRef)
 
-        val dictParams = Array(seg.project, ref.tableAliasName, ref.columnName, seg.kylinconf.getHdfsWorkingDirectory)
+        var dictParams = Array(seg.project, ref.tableAliasName, ref.columnName, seg.kylinconf.getHdfsWorkingDirectory)
           .mkString(SEPARATOR)
         val aliasName = structType.apply(columnIndex).name.concat(ENCODE_SUFFIX)
-        val encodeCol = dict_encode(col(encodeColRef).cast(StringType), lit(dictParams), lit(bucketSize).cast(StringType)).as(aliasName)
-        val columns = partitionedDs.schema.map(ty => col(ty.name)) ++ Seq(encodeCol)
+        var encodeCol = dict_encode(col(encodeColRef).cast(StringType), lit(dictParams), lit(bucketSize).cast(StringType)).as(aliasName)
+        val columns = partitionedDs.schema.map(ty => col(ty.name))
+
+        if (seg.kylinconf.detectDataSkewInDictEncodingEnabled()) {
+          //find skewed data in dict-encoding step
+          val sampleData = ds.sample(seg.kylinconf.sampleRateInEncodingSkewDetection()).cache()
```

Comment:
Prune unused columns before `cache` to reduce memory usage?

Reply:
good suggestion